### PR TITLE
fixes #5848 - Can't search for host with class inherited from config group

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

**Comment by isratrade** [comment]
still need to add unit test

**Comment by domcleal** [comment]
Just a note, it'd be great if you could add to the factory-based tests in test/unit/host_test.rb when adding unit tests.

**Comment by isratrade** [comment]
@domcleal, I added a factory-based test.  However, can you help how I have `FactoryGirl.create(:config_group, :with_puppetclass)` create the puppet class in the environment of the host.
The current `search_by_puppetclass` doesn't take into account in the class in in the host environment. @ohadlevy discussed about caching a hash of the host's classes, but this will be later. I'll add an issue for this feature.

**Comment by domcleal** [comment]
@isratrade have a look at this: https://github.com/domcleal/foreman/commit/c945700148040910ee5d7899db8313e1cc63f762

Now when you create a host group with a config group, it assigns an environment to the host group (that's the environment line within the trait) and passes that into the config group factory as the "class_environment" (the "ignore" works like adding an attr_accessor).  The class_environment is then passed into the puppet class factory so the class is associated to the same environment as the host group.

**Comment by isratrade** [comment]
thanks @domcleal, makes sense. I implement your changes.  My new test passes, but there are 3 failing previous tests with the error:

```
PG::InFailedSqlTransaction: ERROR:  current transaction is aborted, commands ignored until end of transaction block
```

any ideas?

**Comment by domcleal** [comment]
@isratrade strange.. but I can't see any change to the factories that might have caused it.  My suspicion is that something is going wrong when we call `Host.search_for` at test/unit/host_test.rb:925 and it's causing the PostgreSQL transaction to abort.  It's likely that there's an exception handler inside search_for.

Could you try running the test by itself and keeping log/test.log?  It might show underlying errors from PostgreSQL.  Also try looking in PostgreSQL's own log file, as it tends to log them.

File: app/models/concerns/hostext/search.rb [hunk 1]
```diff
@@ -89,15 +89,19 @@ def search_by_user(key, operator, value)
       end
 
       def search_by_puppetclass(key, operator, value)
-        conditions  = sanitize_sql_for_conditions(["puppetclasses.name #{operator} ?", value_to_sql(operator, value)])
-        hosts       = Host.authorized(:view_hosts, Host).where(conditions).joins(:puppetclasses).uniq.map(&:id)
-        host_groups = Hostgroup.unscoped.with_taxonomy_scope.where(conditions).joins(:puppetclasses).uniq.map(&:subtree_ids).flatten.uniq
+        conditions    = sanitize_sql_for_conditions(["puppetclasses.name #{operator} ?", value_to_sql(operator, value)])
+        config_group_ids = ConfigGroup.where(conditions).joins(:puppetclasses).pluck(:id)
+        host_ids         = Host.authorized(:view_hosts, Host).where(conditions).joins(:puppetclasses).uniq.map(&:id)
+        host_ids        += HostConfigGroup.where(:host_type => 'Host::Base').where(:config_group_id => config_group_ids).pluck(:host_id)
```
Comment (domcleal) [inline app/models/concerns/hostext/search.rb:95]: I'm guessing entirely, but I wonder if this (and line 97) cause errors in the transaction if config_group_ids is empty?

File: app/models/concerns/hostext/search.rb [hunk 2]
```diff
@@ -89,15 +89,19 @@ def search_by_user(key, operator, value)
...
 
         opts = ''
-        opts += "hosts.id IN(#{hosts.join(',')})"             unless hosts.blank?
-        opts += " OR "                                        unless hosts.blank? || host_groups.blank?
-        opts += "hostgroups.id IN(#{host_groups.join(',')})"  unless host_groups.blank?
-        opts = "hosts.id < 0"                                 if hosts.blank? && host_groups.blank?
+        opts += "hosts.id IN(#{host_ids.join(',')})"             unless host_ids.blank?
+        opts += " OR "                                        unless host_ids.blank? || hostgroup_ids.blank?
+        opts += "hostgroups.id IN(#{hostgroup_ids.join(',')})"  unless hostgroup_ids.blank?
+        opts = "hosts.id < 0"                                 if host_ids.blank? && hostgroup_ids.blank?
```
Comment (domcleal) [inline app/models/concerns/hostext/search.rb]: Indentation needs tidying on these lines.

File: app/models/concerns/hostext/search.rb (hunk 1 above)
Comment (isratrade) [inline app/models/concerns/hostext/search.rb:95]: @domcleal, yes, your hunch is correct. Thanks!

**Comment by isratrade** [comment]
re-committed with passing tests and fix to indentations

**Comment by domcleal** [comment]
Merged as 81f4af1 for Foreman 1.5.2, thanks @isratrade.

**Pull request closed without merging**